### ISSUE TITLE
Unify storage of handlers in C and Lua

### ISSUE THREAD

**Issue by sdegutis** [sdegutis/mjolnir#241]
### Old notes

Originally titled: "Convert code to use C callbacks and not Apple Blocks"

This is already mostly done in the `hotkey` class on the `next` branch. The one thing left is to filter out numbers from the `hotkey._keys` table in `disableall`. On the bright side, this means we can use `fnutils.filter` instead of `fnutils.copy`, making the latter unused and potentially deletable.

We also need to convert `timer`, `pathwatcher`, and other modules that use callbacks, to use the same style.

---
### New notes

**Motivation:** our current method is tedious, redundant, ugly, and error-prone; and probably overkill for the simple need we have (stopping all event handlers when config is reloaded)

Our event handlers currently have their own `start`, `stop`, and `stopall` methods, named differently for each type. And we need to keep track of them ourselves so that we can stop them automatically from within Hydra when the user reloads their config. That's what [this function](https://github.com/sdegutis/hydra/blob/70e95b13d6a7bdf3989ecebbad16f43b56009d94/Hydra/API/hydra.lua#L7-L15) is doing.

But we're constantly having to reproduce the same code in C which is ugly, tedious, and error-prone. Look in eventtap.m (oops; see below), [hotkey.m](https://github.com/sdegutis/hydra/blob/70e95b13d6a7bdf3989ecebbad16f43b56009d94/Hydra/API/hotkey.m#L96-L101), [pathwatcher.lua](https://github.com/sdegutis/hydra/blob/70e95b13d6a7bdf3989ecebbad16f43b56009d94/Hydra/API/pathwatcher.lua#L25-L26), [timer.lua](https://github.com/sdegutis/hydra/blob/70e95b13d6a7bdf3989ecebbad16f43b56009d94/Hydra/API/timer.lua#L33-L34), [applistener.m](https://github.com/sdegutis/hydra/blob/70e95b13d6a7bdf3989ecebbad16f43b56009d94/Hydra/API/applistener.m#L53-L58), etc.

My proposed solution for this is to create and manage this registry from within ObjC, seeing as we create all these events in C. We could use NSArray (or maybe even an NSIndexSet) to store the ints created from `luaL_ref` for each individual handler. Then we just have to enumerate this array to disable them all when the config reloads. We'd enumerate it from ObjC and call their disable-methods that way. Or better:

We could expose an array inside Lua and some convenience functions in both Lua and C for adding something to it or removing something from it. Each type's own `stop` method defined in Lua would remove self from this array using the Lua convenience function. Each `stop` method defined in C would do the same thing, but via a C convenience function. Each `stopall` method in C would then get a lot easier (the ones written in Lua are already easy) because they could enumerate the array much easier using the convenience function. But wait, that's not actually the hard part. I rewrote [the `applistener.stopall` function in Lua](https://github.com/sdegutis/hydra/blob/70e95b13d6a7bdf3989ecebbad16f43b56009d94/Hydra/API/applistener.lua#L4-L5) and deleted the C version for a different reason: because it was hard to find the right type, get its "stop" field and call it as a function, and somehow it was crashing. Maybe it's not that hard though and I just did something stupid. I don't remember

In conclusion, I'm very confused.
### Related bug report

As a related bug, eventtaps are not currently unregistered when your config reloads. That's a problem, and somewhat related: if it were easier to do, maybe we'd not have forgotten it so easily.
